docs/source/extension/extension_dual_compatibility.rst

```diff
@@ -0,0 +1,107 @@
+written in Typescript).
+
+This is how basic compatibility features work: both apps use the same building
+blocks and methods. For instance, both Lab and Notebook 7 accept Lumino widgets
+as interface components, both apps allow you to add them to the interface by
+specifying an "area" to place them into, and extensions for both use the same
+basic JupyterFrontendPlugin class.
+
+"Do Nothing" Compatibility
+
+This is the easiest way to achieve compatibility: If your extension only uses
+features that both Lab and Notebook 7 have, you can simply install it and it
+will seamlessly work in both Lab and Notebook 7.
+
+An extension that adds a single self-contained icon widget to the top bar
+of the UI, for instance, doesn't need to do anything at all to be compatible
+with both Lab and Notebook 7 (both apps have a top area that can hold the
+widget, so it will be visible in both Lab and Notebook 7 upon install and
+after launch).
+
+Not all extensions will be able to use "Do Nothing" compatibility: The
+"main" area in Lab is one example of a feature/UI area that is not supported
+by Notebook 7 where "Do Nothing" compatibility won't work.
+
+Compatibility in Complex Extensions
+
+Complex extensions can be adapted to work in both apps. The approach and
+effort needed to achieve compatibility depends on the design of your extension.
+
+Some features only exist in Lab, or in Notebook 7, and in those cases you will
+have some extra work to do to ensure compatibility, or to disable features that
+are not available in one of the interfaces. On top of that, you can specialize
+or add behaviors that are specific only to Lab or only to Notebook 7.
+
+With that in mind, here are some general tips for designing compatible extensions:
+
+- Organize your code: Identify which features in your extension are Lab-only
+  and which are Notebook 7-only, and try to separate those from the most
+  important core widgets and classes (you should put this app-specific code
+  in a different place than your main extension classes...if you include
+  app-specific code inside your main classes and widgets, it will be much
+  harder to achieve compatibility and pull those pieces out later)
+- Consider making some features optional, and enable them only when the
+  app-specificfeatures they need are available (more on this "optional-extras"
+  approach later)
+- Note that adding specialized app-specific extra behaviors and features
+  (features that rely on capabilities that only exist in either Lab or
+  Notebook 7) may require more complex compatibility architectures
+
+Solutions
+
+The "Do Nothing" Strategy
+
+Extensions that only use features common to both Lab and Notebook 7 don't
+need any modifications to work in both apps. Users who install your extension
+will be able to seamlessly run it in either Lab or Notebook 7.
+
+Keep in mind that some features for Lab are not available in Notebook 7:
+
+- The "main" area in Lab does not exist in Notebook 7
```

Comment:
Technically there is a `main` area in Notebook 7. But only one widget can be added to it.

We should update that sentence to reflect that.

Reply:
> notification area too

The notification plugin has been added to Notebook 7: https://github.com/jupyter/notebook/pull/6900

But I'm not sure the notification popups are displayed correctly at the moment if an extension uses the notification API. But in theory they should work, at least partially.

Reply:
This section was updated, should be resolved.